File: Week_03/Seminar/Solutions/FirstBadVersion.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left = 0;
        int right = n;
        // you can have additional result variable, but the left index does the same job
        while(left <= right) {
            int mid = left + (right - left) / 2;
            if(isBadVersion(mid)) {
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }
        
        return left;
    }
};

File: Week_04/Practicum/PythonAdditional/First Bad Version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def binary_search(self, left, right):
        while left <= right:
            mid = (left + right) // 2
        
            if isBadVersion(mid):
                right = mid - 1
            else:
                left = mid + 1

        return left
        
    def firstBadVersion(self, n: int) -> int:
        return self.binary_search(0, n)